### PR TITLE
integrate code coverage through codecov.io

### PR DESCRIPTION
**Pull request by Ngalstyan4**
This pull request is meant to integrate code coverage report generation through codecov.io. 
I am not sure this actually works yet but am creating a pull request to trigger travis build and will fix all the issues that arise here.

### CODE REVIEW

**Comment by mfreed** [comment]
Please squash your commits (unless there is a clear organizational reason to have separate), and provide a cleaner summary in the final one(s).  Thanks!

**Comment by Ngalstyan4** [comment]
Instead of hardcoding codecov.io token, shall we create a secure travis environmental variable and read it from there so that we avoid potential misbehaviour? (It takes a simple GET request to push an update to codecov once you have the token)

**Comment by cevian** [comment]
@Ngalstyan4 Do we even need a token? See first q in https://docs.codecov.io/docs/frequently-asked-questions

File: CMakeLists.txt [hunk 1]
```diff
@@ -29,6 +29,29 @@ if (WIN32)
     set(CMAKE_CONFIGURATION_TYPES Release CACHE STRING "Semicolon separated list of supported configuration types, only supports Debug, Release, MinSizeRel, and RelWithDebInfo, anything else will be ignored." FORCE)
   endif ()
 endif (WIN32)
+if ( ENABLE_CODECOVERAGE )
```
Comment (erimatnor) [inline CMakeLists.txt]: Remove whitespaces here and in the rest of the changes since it is not in line with the style in the rest of the CMake file.

**Comment by Ngalstyan4** [comment]
I see. The reason I needed the token was because the build was from inside a docker instance and codecov could not recognize travis. I can now pass travis environment to docker and then we will probably not need the token.

**Pull request merged**